src/looprestoration.rs

```diff
@@ -1779,3 +1792,212 @@ unsafe fn wiener_filter_neon<BD: BitDepthLooprestorationArm>(
+        dav1d_sgr_filter_5x5_8bpc_avx512icl;
+    const dav1d_sgr_filter_3x3_avx512icl: looprestorationfilter_fn =
+        dav1d_sgr_filter_3x3_8bpc_avx512icl;
+    const dav1d_sgr_filter_mix_avx512icl: looprestorationfilter_fn =
+        dav1d_sgr_filter_mix_8bpc_avx512icl;
+}
+
+#[cfg(all(
+    feature = "bitdepth_16",
+    feature = "asm",
+    any(target_arch = "x86", target_arch = "x86_64"),
+))]
+impl BitDepthLoopRestorationAsm for BitDepth16 {
+    // NOTE: There are no SSE2 versions of the wiener filter for 16bpc so we fall back
+    // to the C version.
+    const dav1d_wiener_filter7_sse2: looprestorationfilter_fn = wiener_c_erased::<Self>;
+    const dav1d_wiener_filter5_sse2: looprestorationfilter_fn = wiener_c_erased::<Self>;
+    const dav1d_wiener_filter7_ssse3: looprestorationfilter_fn = dav1d_wiener_filter7_16bpc_ssse3;
+    const dav1d_wiener_filter5_ssse3: looprestorationfilter_fn = dav1d_wiener_filter5_16bpc_ssse3;
+    const dav1d_sgr_filter_5x5_ssse3: looprestorationfilter_fn = dav1d_sgr_filter_5x5_16bpc_ssse3;
+    const dav1d_sgr_filter_3x3_ssse3: looprestorationfilter_fn = dav1d_sgr_filter_3x3_16bpc_ssse3;
+    const dav1d_sgr_filter_mix_ssse3: looprestorationfilter_fn = dav1d_sgr_filter_mix_16bpc_ssse3;
+    const dav1d_wiener_filter7_avx2: looprestorationfilter_fn = dav1d_wiener_filter7_16bpc_avx2;
+    const dav1d_wiener_filter5_avx2: looprestorationfilter_fn = dav1d_wiener_filter5_16bpc_avx2;
+    const dav1d_sgr_filter_5x5_avx2: looprestorationfilter_fn = dav1d_sgr_filter_5x5_16bpc_avx2;
+    const dav1d_sgr_filter_3x3_avx2: looprestorationfilter_fn = dav1d_sgr_filter_3x3_16bpc_avx2;
+    const dav1d_sgr_filter_mix_avx2: looprestorationfilter_fn = dav1d_sgr_filter_mix_16bpc_avx2;
+    const dav1d_wiener_filter7_avx512icl: looprestorationfilter_fn =
+        dav1d_wiener_filter7_16bpc_avx512icl;
+    const dav1d_wiener_filter5_avx512icl: looprestorationfilter_fn =
+        dav1d_wiener_filter5_16bpc_avx512icl;
+    const dav1d_sgr_filter_5x5_avx512icl: looprestorationfilter_fn =
+        dav1d_sgr_filter_5x5_16bpc_avx512icl;
+    const dav1d_sgr_filter_3x3_avx512icl: looprestorationfilter_fn =
+        dav1d_sgr_filter_3x3_16bpc_avx512icl;
+    const dav1d_sgr_filter_mix_avx512icl: looprestorationfilter_fn =
+        dav1d_sgr_filter_mix_16bpc_avx512icl;
+}
+
+#[cfg(all(feature = "asm", any(target_arch = "x86", target_arch = "x86_64")))]
+#[inline(always)]
+fn loop_restoration_dsp_init_x86<BD: BitDepthLoopRestorationAsm>(
+    c: &mut Dav1dLoopRestorationDSPContext,
+    bpc: libc::c_int,
+) {
+    // TODO(randomPoison): Import temporarily needed until init fns are deduplicated.
+    use crate::src::looprestoration::*;
+    use crate::src::x86::cpu::*;
+
+    let flags = dav1d_get_cpu_flags();
+
+    if flags & DAV1D_X86_CPU_FLAG_SSE2 == 0 {
+        return;
+    }
+
+    c.wiener[0] = BD::dav1d_wiener_filter7_sse2;
+    c.wiener[1] = BD::dav1d_wiener_filter5_sse2;
```

Comment:
With the `match` way in #355, implementing this like C does and without making it confusing by setting the sse2 version to the C version, this becomes very simple.